src/main/java/com/openlattice/datastore/data/controllers/DataController.java

```diff
@@ -414,6 +418,82 @@ public Void clearEntitySet( UUID entitySetId ) {
         return null;
     }
 
+    @Override
+    @Timed
+    @RequestMapping(
+            path = { "/" + ENTITY_SET + "/" + SET_ID_PATH + "/" + ENTITY_KEY_ID_PATH + "/" + NEIGHBORS },
+            method = RequestMethod.DELETE
+    )
+    public Long clearEntityAndNeighborEntities(
+            @PathVariable( ENTITY_SET_ID ) UUID vertexEntitySetId,
+            @PathVariable( ENTITY_KEY_ID ) UUID vertexEntityKeyId
+    ) {
+
+        /*
+         * 1 - collect all relevant EntitySets and PropertyTypes
+         */
+
+        // getNeighborEntitySetIds() returns source, destination, and edge EntitySet ids
+        Set<UUID> neighborEntitySetIds = dgm.getNeighborEntitySetIds( vertexEntitySetId );
+        Set<UUID> allEntitySetIds = ImmutableSet.<UUID>builder()
+                .add( vertexEntitySetId )
+                .addAll( neighborEntitySetIds )
+                .build();
+
+        Map<AclKey, EnumSet<Permission>> entitySetsAccessRequestMap = allEntitySetIds
+                .stream()
+                .map( AclKey::new )
+                .collect( Collectors.toMap( Function.identity(), aclKey -> READ_PERMISSION ) );
+
+        Map<UUID, Map<UUID, PropertyType>> entitySetIdToPropertyTypesMap = Maps.newHashMap();
+        Map<AclKey, EnumSet<Permission>> propertyTypesAccessRequestMap = allEntitySetIds
+                .stream()
+                .flatMap( esId -> {
+                    Map<UUID, PropertyType> propertyTypes = edmService.getPropertyTypesForEntitySet( esId );
+                    entitySetIdToPropertyTypesMap.put( esId, propertyTypes );
+                    return entitySetIdToPropertyTypesMap.keySet().stream().map( ptId -> new AclKey( esId, ptId ) );
+                } )
+                .collect( Collectors.toMap( Function.identity(), aclKey -> WRITE_PERMISSION ) );
+
+        /*
+         * 2 - check permissions for all relevant EntitySets and PropertyTypes
+         */
+
+        Map<AclKey, EnumSet<Permission>> accessRequestMap = Maps.newHashMap();
```

Comment:
Unnecessary double copy.

Reply:
accessRequestMap = entitySetsAccessRequestMap + propertyTypesAccessRequestMap

You can avoid a lot of the work here by ```Stream<AclKey> entitySetsAccessRequests = ...``` and ```Stream<AclKey> propertyTypeAccessRequests = ...``` and then ```Stream.concat(entitySetsAccessRequests, propertyTypeAccessRequest).collect( Function.identity(), aclKey -> aclKey.size() == 1 ? READ_PERMISSION : WRITE_PERMISSION)```